### PR TITLE
Implement Refresh Worker Certificates Logic

### PR DESCRIPTION
**Pull request by mateoflorido**
> [!IMPORTANT]  
> This pull request is dependent on `k8s-snap` [#713](https://github.com/canonical/k8s-snap/pull/713)
## Overview
This update implements the logic to handle certificate refresh for worker nodes.
## Changes
- Added functionality to refresh certificates for worker nodes.
- Refactored the Workload Cluster API to support certificate refreshing.

### CODE REVIEW

File: pkg/ck8s/workload_cluster.go [hunk 1]
```diff
@@ -242,7 +243,32 @@ func (w *Workload) GetCertificatesExpiryDate(ctx context.Context, machine *clust
...
+func (w *Workload) ApproveCertificates(ctx context.Context, machine *clusterv1.Machine, capiToken string, seed int) error {
+	request := ApproveWorkerCSRRequest{}
+	response := &ApproveWorkerCSRResponse{}
+	k8sdProxy, err := w.GetK8sdProxyForControlPlane(ctx, k8sdProxyOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create k8sd proxy: %w", err)
+	}
+
+	header := map[string][]string{
+		"capi-auth-token": {w.authToken},
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:261]: I think it might not be a bad idea to either extract this string as a const or have a helper function like `NewHeaderWithAuthToken`. Maybe something for a future improvement tho.

File: pkg/ck8s/workload_cluster.go [hunk 2]
```diff
@@ -259,17 +285,76 @@ func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.M
...
+	header := map[string][]string{
+		"node-token": {nodeToken},
+	}
+
+	k8sdProxy, err := w.GetK8sdProxyForMachine(ctx, machine)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create k8sd proxy: %w", err)
+	}
+
+	if err := w.doK8sdRequest(ctx, k8sdProxy, http.MethodPost, "1.0/x/capi/refresh-certs/run", header, request, runResponse); err != nil {
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:302]: Maybe we need to replace this with: https://github.com/canonical/k8s-snap-api/blob/main/api/v1/rpc_clusterapi_certificates_run.go#L3

File: pkg/ck8s/workload_cluster.go [hunk 3]
```diff
@@ -242,7 +243,32 @@ func (w *Workload) GetCertificatesExpiryDate(ctx context.Context, machine *clust
...
 
-func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.Machine, nodeToken string, expirationSeconds int, extraSANs []string) (int, error) {
+type ApproveWorkerCSRRequest struct {
+	Seed int `json:"seed"`
+}
+
+type ApproveWorkerCSRResponse struct{}
+
+func (w *Workload) ApproveCertificates(ctx context.Context, machine *clusterv1.Machine, capiToken string, seed int) error {
+	request := ApproveWorkerCSRRequest{}
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:253]: Just a reminder to replace this once the api is merged.

File: pkg/ck8s/workload_cluster.go [hunk 4]
```diff
@@ -242,7 +243,32 @@ func (w *Workload) GetCertificatesExpiryDate(ctx context.Context, machine *clust
...
+	k8sdProxy, err := w.GetK8sdProxyForControlPlane(ctx, k8sdProxyOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create k8sd proxy: %w", err)
+	}
+
+	header := map[string][]string{
+		"capi-auth-token": {w.authToken},
+	}
+
+	if err := w.doK8sdRequest(ctx, k8sdProxy, http.MethodPost, "1.0/x/capi/refresh-certs/approve", header, request, response); err != nil {
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:264]: Just a reminder to replace this once the api is merged.

File: pkg/ck8s/workload_cluster.go [hunk 5]
```diff
@@ -259,17 +285,76 @@ func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.M
...
+		Seed:              seed,
+		ExpirationSeconds: expirationSeconds,
+	}
+
+	var seconds int
+
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		seconds, err = w.refreshCertificatesRun(ctx, machine, nodeToken, &request)
+		return err
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:325]: Maybe we can wrap this?

File: pkg/ck8s/workload_cluster.go [hunk 6]
```diff
@@ -259,17 +285,76 @@ func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.M
...
+	})
+
+	eg.Go(func() error {
+		return w.ApproveCertificates(ctx, machine, nodeToken, seed)
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:329]: Maybe we can wrap this?

**Review by HomayoonAlimohammadi — COMMENTED** [review]
Thanks a lot @mateoflorido! Wonderful job! Almost LGTM, just left some minor comments and questions.

File: pkg/ck8s/workload_cluster.go [hunk 7]
```diff
@@ -259,17 +285,81 @@ func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.M
...
+	}
+
+	if err := w.doK8sdRequest(ctx, k8sdProxy, http.MethodPost, "1.0/x/capi/refresh-certs/run", header, request, runResponse); err != nil {
+		return 0, fmt.Errorf("failed to run refresh certificates: %w", err)
+	}
+
+	return runResponse.ExpirationSeconds, nil
+}
+
+func (w *Workload) RefreshWorkerCertificates(ctx context.Context, machine *clusterv1.Machine, nodeToken string, expirationSeconds int) (int, error) {
```
Comment (berkayoz) [inline pkg/ck8s/workload_cluster.go:309]: Can we add a comment here explaining how this flow works? What are the steps etc.

**Review by berkayoz — COMMENTED** [review]
Great work, did an initial pass. We should document/comment well here since the flow is a bit more complex than normal.

File: pkg/ck8s/workload_cluster.go (hunk 2 above)
Comment (mateoflorido) [reply pkg/ck8s/workload_cluster.go:270]: Unfortunately, we cannot replace the path here with the k8s-snap-api const, because it's missing the version prefix. I wonder if using something like string concatenation would be optimized by the compiler.

Comment (mateoflorido) [reply pkg/ck8s/workload_cluster.go:270]: It's seems that the compiler optimizes this behavior. I'll create a follow up PR to address this.

File: pkg/ck8s/workload_cluster.go [hunk 8]
```diff
@@ -209,13 +208,28 @@ func (w *Workload) GetCertificatesExpiryDate(ctx context.Context, machine *clust
...
+	request := apiv1.ClusterAPIApproveWorkerCSRRequest{}
+	response := &apiv1.ClusterAPIApproveWorkerCSRResponse{}
+	k8sdProxy, err := w.GetK8sdProxyForControlPlane(ctx, k8sdProxyOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create k8sd proxy: %w", err)
+	}
+
+	header := w.newHeaderWithCAPIAuthToken()
+
+	if err := w.doK8sdRequest(ctx, k8sdProxy, http.MethodPost, "1.0/x/capi/refresh-certs/approve", header, request, response); err != nil {
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:221]: Maybe we can use the `k8s-snap-api` endpoint here? Something like `url.JoinPath(apiv1.K8sdAPIVersion, apiv1.ClusterAPICertificatesApproveRPC)`

File: pkg/ck8s/workload_cluster.go [hunk 9]
```diff
@@ -226,17 +240,91 @@ func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.M
...
+func (w *Workload) refreshCertificatesRun(ctx context.Context, machine *clusterv1.Machine, nodeToken string, request *apiv1.ClusterAPICertificatesRunRequest) (int, error) {
+	runResponse := &apiv1.ClusterAPICertificatesRunResponse{}
+	header := w.newHeaderWithNodeToken(nodeToken)
+
+	k8sdProxy, err := w.GetK8sdProxyForMachine(ctx, machine)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create k8sd proxy: %w", err)
+	}
+
+	if err := w.doK8sdRequest(ctx, k8sdProxy, http.MethodPost, "1.0/x/capi/refresh-certs/run", header, request, runResponse); err != nil {
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:255]: Maybe we can also use `k8s-snap-api` here as well?

File: pkg/ck8s/workload_cluster.go [hunk 10]
```diff
@@ -226,17 +240,91 @@ func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.M
...
+
+	return runResponse.ExpirationSeconds, nil
+}
+
+// RefreshWorkerCertificates approves the worker node CSR and refreshes the certificates.
+// The certificate approval process follows these steps:
+// 1. The CAPI provider calls the /x/capi/refresh-certs/plan endpoint from the
+// worker node, which generates the CSRs and creates the CertificateSigningRequest
+// objects in the cluster.
+// 2. The CAPI provider then calls the /x/capi/refresh-certs/plan endpoint with
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:267]: _I think_ this should be `/run`

File: pkg/ck8s/workload_cluster.go [hunk 11]
```diff
@@ -226,17 +240,91 @@ func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.M
...
+// the seed. This endpoint waits until the CSR is approved and the certificate
+// is signed. Note that this is a blocking call.
+// 3. The CAPI provider calls the /x/capi/refresh-certs/approve endpoint from
+// any control plane node to approve the CSRs.
+// 4. The /x/capi/refresh-certs/plan endpoint completes and returns once the
```
Comment (HomayoonAlimohammadi) [inline pkg/ck8s/workload_cluster.go:272]: _I think_ this should be `/run` as well

**Review by HomayoonAlimohammadi — APPROVED** [review]
Amazing job! Thanks a lot @mateoflorido! LGTM, Just some very minor comments.

File: go.mod [hunk 12]
```diff
@@ -3,7 +3,7 @@ module github.com/canonical/cluster-api-k8s
 go 1.22.6
 
 require (
-	github.com/canonical/k8s-snap-api v1.0.9
+	github.com/canonical/k8s-snap-api v1.0.10
```
Comment (berkayoz) [inline go.mod:6]: We should be careful here as we have another PR to bump this to `v1.0.11`.
#72

**Review by berkayoz — APPROVED** [review]
LGTM, seems like @HomayoonAlimohammadi covered most of the comments.

File: pkg/ck8s/workload_cluster.go (hunk 8 above)
Comment (mateoflorido) [reply pkg/ck8s/workload_cluster.go:221]: Let's create a follow up PR to refactor the URLs across the codebase. It would be difficult to introduce this type of change in some parts while leaving others unchanged. What do you think?

File: pkg/ck8s/workload_cluster.go (hunk 9 above)
Comment (mateoflorido) [reply pkg/ck8s/workload_cluster.go:255]: https://github.com/canonical/cluster-api-k8s/pull/65#discussion_r1805602467

File: pkg/ck8s/workload_cluster.go [hunk 13]
```diff
@@ -226,17 +240,91 @@ func (w *Workload) RefreshCertificates(ctx context.Context, machine *clusterv1.M
...
+// certificate is approved and signed.
```
[the remark this answers is not in the document]
Comment (mateoflorido) [reply pkg/ck8s/workload_cluster.go:273]: Sure, let's create a follow-up card

File: pkg/ck8s/workload_cluster.go (hunk 8 above)
Comment (HomayoonAlimohammadi) [reply pkg/ck8s/workload_cluster.go:221]: Of course, let's go!

**Comment by bschimke95** [comment]
LGTM @mateoflorido but as also discussed yesterday with @HomayoonAlimohammadi I'd like to have a test for this before we merge this.

**Review by bschimke95 — APPROVED** [review]
LGTM, amazing work @mateoflorido
